src/compiler/utilities.ts

```diff
@@ -1904,6 +1904,76 @@ namespace ts {
+            // removeComments is true, only reserve pinned comment at the top of file
+            // For example:
+            //      /*! Pinned Comment */
+            //
+            //      var x = 10;
+            if (node.pos === 0) {
+                leadingComments = filter(getLeadingCommentRanges(currentSourceFile.text, node.pos), isPinnedComments);
+            }
+        }
+        else {
+            // removeComments is false, just get detached as normal and bypass the process to filter comment
+            leadingComments = getLeadingCommentRanges(currentSourceFile.text, node.pos);
+        }
+
+        if (leadingComments) {
+            let detachedComments: CommentRange[] = [];
+            let lastComment: CommentRange;
+
+            forEach(leadingComments, comment => {
+                if (lastComment) {
+                    let lastCommentLine = getLineOfLocalPosition(currentSourceFile, lastComment.end);
+                    let commentLine = getLineOfLocalPosition(currentSourceFile, comment.pos);
+
+                    if (commentLine >= lastCommentLine + 2) {
+                        // There was a blank line between the last comment and this comment.  This
+                        // comment is not part of the copyright comments.  Return what we have so
+                        // far.
+                        return detachedComments;
+                    }
+                }
+
+                detachedComments.push(comment);
+                lastComment = comment;
+            });
+
+            if (detachedComments.length) {
+                // All comments look like they could have been part of the copyright header.  Make
+                // sure there is at least one blank line between it and the node.  If not, it's not
+                // a copyright header.
+                let lastCommentLine = getLineOfLocalPosition(currentSourceFile, lastOrUndefined(detachedComments).end);
+                let nodeLine = getLineOfLocalPosition(currentSourceFile, skipTrivia(currentSourceFile.text, node.pos));
+                if (nodeLine >= lastCommentLine + 2) {
+                    // Valid detachedComments
+                    emitNewLineBeforeLeadingComments(currentSourceFile, writer, node, leadingComments);
+                    emitComments(currentSourceFile, writer, detachedComments, /*trailingSeparator*/ true, newLine, writeComment);
+                    let currentDetachedCommentInfo = { nodePos: node.pos, detachedCommentEndPos: lastOrUndefined(detachedComments).end };
+                    if (detachedCommentsInfo) {
+                        detachedCommentsInfo.push(currentDetachedCommentInfo);
+                    }
+                    else {
+                        detachedCommentsInfo = [currentDetachedCommentInfo];
+                    }
+                }
+            }
+
+            return detachedCommentsInfo;
+        }
+
+        function isPinnedComments(comment: CommentRange) {

```

Comment:
There is already an `isPinnedComments` elsewhere. Also, can you rename it to `isPinnedComment`?